FirebaseStorage/Sources/StorageMetadata.swift

```diff
@@ -162,80 +105,199 @@ import FirebaseStorageInternal
+    }
+    if let md5Hash = md5Hash {
+      dictionary["md5Hash"] = md5Hash
+    }
+    if let customMetadata = customMetadata {
+      dictionary["metadata"] = customMetadata
+    }
+    if size != 0 {
+      dictionary["size"] = size
+    }
+    if generation != 0 {
+      dictionary["generation"] = "\(generation)"
+    }
+    if metageneration != 0 {
+      dictionary["metageneration"] = "\(metageneration)"
+    }
+    if let timeCreated = timeCreated {
+      dictionary["timeCreated"] = StorageMetadata.RFC3339StringFromDate(timeCreated)
+    }
+    if let updated = updated {
+      dictionary["updated"] = StorageMetadata.RFC3339StringFromDate(updated)
+    }
+    if let path = path {
+      dictionary["name"] = path
+    }
+    return dictionary
   }
 
   /**
    * Determines if the current metadata represents a "file".
    */
-  @objc public let isFile: Bool
+  @objc public var isFile: Bool {
+    return fileType == .file
+  }
 
+  // TODO: Nothing in the implementation ever sets `.folder`.
   /**
    * Determines if the current metadata represents a "folder".
    */
-  @objc public let isFolder: Bool
+  @objc public var isFolder: Bool {
+    return fileType == .folder
+  }
 
   // MARK: - Public Initializers
 
   @objc override public convenience init() {
-    self.init(impl: FIRIMPLStorageMetadata(dictionary: [:])!)
+    self.init(dictionary: [:])
   }
 
   /**
    * Creates an instance of StorageMetadata from the contents of a dictionary.
    * @return An instance of StorageMetadata that represents the contents of a dictionary.
    */
-  @objc public convenience init(dictionary: [String: Any]) {
-    self.init(impl: FIRIMPLStorageMetadata(dictionary: dictionary)!)
+  @objc public init(dictionary: [String: AnyHashable]) {
+    initialMetadata = dictionary
+    bucket = dictionary["bucket"] as? String ?? nil
+    cacheControl = dictionary["cacheControl"] as? String ?? nil
+    contentDisposition = dictionary["contentDisposition"] as? String ?? nil
+    contentEncoding = dictionary["contentEncoding"] as? String ?? nil
+    contentLanguage = dictionary["contentLanguage"] as? String ?? nil
+    contentType = dictionary["contentType"] as? String ?? nil
+    customMetadata = dictionary["metadata"] as? [String: String] ?? nil
+    size = StorageMetadata.intFromDictionaryValue(dictionary["size"])
+    generation = StorageMetadata.intFromDictionaryValue(dictionary["generation"])
+    metageneration = StorageMetadata.intFromDictionaryValue(dictionary["metageneration"])
+    timeCreated = StorageMetadata.dateFromRFC3339String(dictionary["timeCreated"])
+    updated = StorageMetadata.dateFromRFC3339String(dictionary["updated"])
+    md5Hash = dictionary["md5Hash"] as? String ?? nil
+
+    // GCS "name" is our path, our "name" is just the last path component of the path
+    if let name = dictionary["name"] as? String {
+      path = name
+      self.name = (name as NSString).lastPathComponent
+    }
+    fileType = .unknown
```

Comment:
Probably out of scope for this PR, but this and the `dictionaryRepresentation` method above suggest this class should conform to Codable (though the former methods still need to remain available to objc users).

Reply:
Good point. Agreed out of scope for this pr.